### PR TITLE
Add tests for trace_api.

### PR DESCRIPTION
**Pull request by faramir-dev**
Fixes openssl#17422

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/trace.c [hunk 1]
```diff
@@ -146,19 +146,33 @@ const char *OSSL_trace_get_category_name(int num)
 {
     size_t i;
 
+    /*
+     * Partial check that OSSL_TRACE_CATEGORY_... macros
+     * are synced with trace_categories array
+     */
+    static const size_t trace_categories_len =
+        sizeof(trace_categories)/sizeof(0[trace_categories]);
```
Comment (levitte) [inline crypto/trace.c:154]: There's an OpenSSL-specific macro `OSSL_NELEM()` to get the amount of elements in an array.  You can see it used in the `for` loop a few lines down.

Also, hehehe, it's not often that I see the construct `n[pointer]` used.  Yes, it is legitimate C, but so unusual that it's not what would be called reader friendly code.  While I can appreciate the coolness factor, I'd suggest not doing this in OpenSSL source code, yeah?

File: crypto/trace.c [hunk 2]
```diff
@@ -146,19 +146,32 @@ const char *OSSL_trace_get_category_name(int num)
 {
     size_t i;
 
+    /*
+     * Partial check that OSSL_TRACE_CATEGORY_... macros
+     * are synced with trace_categories array
+     */
+    static const size_t trace_categories_len = OSSL_NELEM(trace_categories);
+    if (!ossl_assert(OSSL_TRACE_CATEGORY_NUM == trace_categories_len))
```
Comment (t8m) [inline crypto/trace.c:154]: Please add empty line above (and probably drop the one after the declaration of i).

File: crypto/trace.c [hunk 3]
```diff
@@ -146,19 +146,32 @@ const char *OSSL_trace_get_category_name(int num)
...
+        return NULL;
+
     for (i = 0; i < OSSL_NELEM(trace_categories); i++)
```
Comment (t8m) [inline crypto/trace.c:157]: Please use `trace_categories_len` here.

File: test/trace_api_test.c [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/trace.h>
+
+#include "testutil.h"
+
+static int test_trace_categories(void)
+{
+    for (int cat_num = -1; cat_num <= OSSL_TRACE_CATEGORY_NUM + 1; ++cat_num) {
```
Comment (t8m) [inline test/trace_api_test.c:16]: Do not declare variables in the middle of a block.

File: test/trace_api_test.c [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        default:
+            is_cat_name_eq = TEST_ptr_null(cat_name);
+            break;
+        }
+
+        if (!TEST_true(is_cat_name_eq))
+            return 0;
+        const int ret_cat_num =
+            OSSL_trace_get_category_num(cat_name);
+        const int expected_ret = cat_name != NULL ? cat_num : -1;
```
Comment (t8m) [inline test/trace_api_test.c:55]: As above about the declarations.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/trace.c (hunk 3 above)
Comment (levitte) [reply crypto/trace.c:159]: The other option is to drop `trace_categories_len` and replace `OSSL_TRACE_CATEGORY_NUM == trace_categories_len` with `OSSL_TRACE_CATEGORY_NUM == OSSL_NELEM(trace_categories)` above.

File: test/trace_api_test.c [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+ */
+
+#include <openssl/trace.h>
+
+#include "testutil.h"
+
+static int test_trace_categories(void)
+{
+    int cat_num;
+    for (cat_num = -1; cat_num <= OSSL_TRACE_CATEGORY_NUM + 1; ++cat_num) {
```
Comment (t8m) [inline test/trace_api_test.c:17]: Missing empty line after declaration.

**Review by t8m — CHANGES_REQUESTED** [review]
One last nit

**Review by t8m — APPROVED** [review]

File: test/trace_api_test.c [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        const char *cat_name = OSSL_trace_get_category_name(cat_num);
+        int is_cat_name_eq = 0;
+        int ret_cat_num;
+        int expected_ret;
+
+        switch (cat_num) {
+#define CASE(name) \
+        case OSSL_TRACE_CATEGORY_##name: \
+            is_cat_name_eq = TEST_str_eq(cat_name, #name); \
+            break;
```
Comment (levitte) [inline test/trace_api_test.c:28]: A recommendation: drop the final `;`.  That allows using using the macro in a way that source code formaters can handle a little better (yup, the `;` makes a difference in that sense):

``` C
            CASE(TRACE);
```

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
@t8m, if you approve the latest fixup, we're done

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you.

**Pull request closed without merging**